rfcs/react-components/convergence/motion-definition-n-apis.md

```diff
@@ -0,0 +1,1079 @@
+  // Heads up! `duration` is in milliseconds and `easing` is a string
+  // ⚠️ We can't use CSS variables there
+  duration: motionTokens.durationSlow,
+  easing: motionTokens.accelerateMax,
+};
+```
+
+Does this mean that CSS variables can't be used for them at all? **No**, as motions can also be defined as factories that accept an animated element as an argument:
+
+```tsx
+// ⚠️ This is not proposed API, it's just an example
+
+const motion: MotionAtomFn = element => {
+  const computedStyle = getComputedStyle(element);
+
+  return {
+    duration: Number(computedStyle.getPropertyValue('--durationUltraSlow').replace('ms', '')),
+    easing: computedStyle.getPropertyValue('--curveAccelerateMax'),
+  };
+};
+```
+
+While this approach should not be commonly used, it's worth noting that it's possible.
+
+[An example on StackBlitz](https://stackblitz.com/edit/stackblitz-starters-eqp8gv)
+
+</details>
+
+### Using atom motions
+
+The proposal is to expose factories as they provide more flexibility and can be used with custom motion definitions.
+
+```tsx
+import { createAtom, atoms } from '@fluentui/react-motions-preview';
+
+const FadeEnterSlow = createAtom(atoms.fade.enterSlow);
+
+function MyComponent() {
+  return (
+    <FadeEnterSlow iterations={Infinity}>
+      <div>Hello world!</div>
+    </FadeEnterSlow>
+  );
+}
+```
+
+> Note: `createAtom()` returns a React component that clones the child element and applies the provided animation to it.
+
+<details>
+  <summary>Optional utility components?</summary>
+
+We can also expose a set of pre-defined components, for example:
+
+```tsx
+import { Collapse } from '@fluentui/react-motions-preview';
+
+function MyComponent() {
+  return (
+    <Collapse>
+      <div>Hello world!</div>
+    </Collapse>
+  );
+}
+```
+
+However, as we need to expose factories anyway, the proposal is to expose them as the primary API.
+
+</details>
+
+<details>
+  <summary>Why not follow `styled-components` / `framer-motion` API?</summary>
+
+Another API approach might be to compose existing components with motion factories, for example:
+
+```tsx
+// ⚠️ This is not proposed API, it's just an example
+
+import { Button } from '@fluentui/react-components';
+import { createAtomElement, atom } from '@fluentui/react-motions-preview';
+
+const FadeEnterDiv = createAtomElement('div', atom.fade.enterSlow);
+const FadeEnterButton = createAtomElement(Button, atom.fade.enterSlow);
+
+function MyComponent() {
+  return (
+    <>
+      <FadeEnterDiv>Hello world!</FadeEnterDiv>
+      <FadeEnterButton>Hello world!</FadeEnterButton>
+    </>
+  );
+}
+```
+
+This approach does not have obvious pros or cons, but it's worth noting that every animated component will have to be composed, for example:
+
+```tsx
+// ⚠️ This is not proposed API, it's just an example
+
+import { Button, Image } from '@fluentui/react-components';
+import { createAtom, createAtomElement, atom } from '@fluentui/react-motions-preview';
```

Comment:
This is a neat API for creating reusable animation components based on new or existing atoms, but the naming here might conflict with well-established state/store libraries in the React ecosystem, like Recoil/Jotai, and possibly other libraries that uses the "atom" concept. 

Working with the concept of atoms does not necessarily mean that we need to name them "atom" and "createAtom".

If a project uses two libraries that implement "atoms", it might be confusing:

``` ts
import { createAtom, atom: motionAtom } from '@fluentui/react-motions-preview';
import { atom } from 'recoil'

// ... code

// Both are "global" scoped
const FadeEnter = createAtom(motionAtom.fade.enterSlow);
const todoListState = atom({
  key: 'TodoList',
  default: [],
});

// ... code
```

Even though the import comes from `@fluentui/react-motions` or `@fluentui/react-motion` and they can be renamed to something else, it might still cause confusion.

Could we evaluate if it is worth renaming it to something else, more scoped?

Something like: 

``` ts
import { createMotion, motion } from '@fluentui/react-motions-preview';

const FadeEnter = createMotion(motion.fade.enterSlow);
```

Or even:

``` ts
import { createMotionAtom, motionAtom } from '@fluentui/react-motions-preview';

const FadeEnter = createMotionAtom(motionAtom.fade.enterSlow);
```

Reply:
I just realized this comment was misplaced, as it was not a proposed API but an example. But it is relevant for the whole API naming convention, so I'll keep it here.

Reply:
I think that it's a good point. What about the following?

- `createAtom()` => `createMotionAtom()`
- `createPresence()` => `createMotionPresence()`

/cc @robertpenner

Reply:
@marcosmoura I updated to RFC to use `createAtomMotion()` & `createPresenceMotion()`